tests/utils/variable_consistency_checker.py

```diff
@@ -0,0 +1,217 @@
+            glossary[keys[0]] = description
+
+        return glossary, set(glossary.keys())
+
+    def read_algorithm_keys(self) -> set[str]:
+        """Read the algorithm keys."""
+        algorithm_keys: set[str] = set()
+        for alg in Algorithm.instances.values():
+            algorithm_keys.update(alg.input_keys)
+            algorithm_keys.update(alg.return_keys)
+
+        return algorithm_keys
+
+    def read_variables_dict(self) -> set[str]:
+        """Read the variables dictionary file."""
+        with as_file(files("cfspopcon").joinpath("variables.yaml")) as filepath:
+            with open(filepath) as f:
+                variables_dict = yaml.safe_load(f)
+
+        return variables_dict, set(variables_dict.keys())
+
+    def run(self, apply_changes: bool = True) -> None:
+        """Check the files and, if apply_changes = True, modify the files in place."""
+        success = True
+
+        unused_variable_keys = self.variable_keys - self.algorithm_keys
+        unlisted_args = self.algorithm_keys - self.variable_keys
+
+        extra_glossary_keys = (self.glossary_keys - self.variable_keys) - unlisted_args
+        undocumented_args = (self.variable_keys - self.glossary_keys) - unused_variable_keys
+
+        linebreak = "\n"
+        if len(unused_variable_keys):
+            print(
+                f"{linebreak}The following keys in the variables dictionary are not used by any algorithm and will be removed:{linebreak}{linebreak.join(list(unused_variable_keys))}{linebreak}"
+            )
+
+        if len(unlisted_args):
+            print(
+                f"{linebreak}The following Algorithm input/output keys are not defined in the variables dictionary and will be added:{linebreak}{linebreak.join(list(unlisted_args))}{linebreak}"
+            )
+
+        if len(extra_glossary_keys):
+            print(
+                f"{linebreak}The following keys in the glossary are not in the variables dictionary and will be removed:{linebreak}{linebreak.join(list(extra_glossary_keys))}{linebreak}"
+            )
+
+        if len(undocumented_args):
+            print(
+                f"{linebreak}The following keys in the variables dictionary are not defined in the glossary and will be added:{linebreak}{linebreak.join(list(undocumented_args))}{linebreak}"
+            )
+
+        if len(unused_variable_keys) or len(unlisted_args) or len(extra_glossary_keys) or len(undocumented_args):
+            success = False
+
+        if not apply_changes:
+            exit(0) if success else exit(1)
+
+        new_variables_dict = dict()
+        for key in sorted((self.variable_keys - unused_variable_keys) | unlisted_args, key=str.lower):
+            used_by = []
+            set_by = []
+
+            for alg in Algorithm.instances.values():
+                if key in alg.input_keys:
+                    used_by.append(alg._name)
+                if key in alg.return_keys:
+                    set_by.append(alg._name)
```

Comment:
could consider moving this out of the loop and traversing the algorithms list once to create a dict of key -> (used_by, set_by). Would avoid having to do this full loop for every key.